jgrapht-core/src/main/java/org/jgrapht/alg/matching/blossom/v5/KolmogorovMinimumWeightPerfectMatching.java

```diff
@@ -0,0 +1,889 @@
+ * a mapping from the node sets of odd cardinality to the corresponding dual variables. This mapping doesn't contain
+ * the sets whose dual variables are $0$. The computation of the dual solution is performed lazily and doesn't affect
+ * the running time of finding a minimum weight perfect matching.
+ * <p>
+ * This class allows to test the optimality of the solution via {@link KolmogorovMinimumWeightPerfectMatching#testOptimality()}.
+ * It also allows to get the error of the computation when the edge weights are real values via
+ * {@link KolmogorovMinimumWeightPerfectMatching#getError()}. Both optimality test and error computation are performed
+ * lazily and don't affect the running time of the main algorithm. If the problem instance doesn't contain a perfect
+ * matching at all, the algorithm doesn't find a minimum weight maximum matching, it thrown an exception.
+ *
+ * @param <V> the graph vertex type
+ * @param <E> the graph edge type
+ * @author Timofey Chudakov
+ * @see BlossomVPrimalUpdater
+ * @see BlossomVDualUpdater
+ * @since June 2018
+ */
+public class KolmogorovMinimumWeightPerfectMatching<V, E> implements MatchingAlgorithm<V, E> {
+    /**
+     * Default epsilon used in the algorithm
+     */
+    public static final double EPS = MatchingAlgorithm.DEFAULT_EPSILON;
+    /**
+     * Default infinity value used in the algorithm
+     */
+    public static final double INFINITY = 1e100;
+    /**
+     * Defines the threshold for throwing an exception about no perfect matching existence
+     */
+    public static final double NO_PERFECT_MATCHING_THRESHOLD = 1e10;
+    /**
+     * Variable for debug purposes
+     */
+    static final boolean DEBUG = false;
+    /**
+     * Message about no perfect matching
+     */
+    static final String NO_PERFECT_MATCHING = "There is no perfect matching in the specified graph";
+    /**
+     * Default options
+     */
+    private static final BlossomVOptions DEFAULT_OPTIONS = new BlossomVOptions();
+
+    /**
+     * The graph we are matching on
+     */
+    private final Graph<V, E> graph;
+
+    /**
+     * Current state of the algorithm
+     */
+    BlossomVState<V, E> state;
+    /**
+     * Perform primal operations (grow, augment, shrink and expand)
+     */
+    private BlossomVPrimalUpdater<V, E> primalUpdater;
+    /**
+     * Performs dual updates using the strategy defined by the {@code options}
+     */
+    private BlossomVDualUpdater<V, E> dualUpdater;
+    /**
+     * The computed matching of the {@code graph}
+     */
+    private MatchingAlgorithm.Matching<V, E> matching;
+    /**
+     * Defines solution to the dual linear program formulated on the {@code graph}
+     */
+    private DualSolution dualSolution;
+    /**
+     * BlossomVOptions used by the algorithm to getMatching the problem instance
+     */
+    private BlossomVOptions options;
+
+    /**
+     * Constructs a new instance of the algorithm with the default options for it.
```

Comment:
Should we use GraphType to validate that the input graph is indeed weighted?  Your comments mention that for unweighted graphs, it is better to use one of our other available algorithms, but perhaps there are still cases where this algorithm would be optimal?

Reply:
We can use GraphType for this purpose, but should we throw an exception in the case the graph isn't weighted? This doesn't seem to be an error. 

This is really a good question. I'll compare Blossom V with EdmondsMaximumCardinalityMatching. I think at least memory usage will be much higher.

Reply:
One way to implement a "warning" for this is to make it an exception by default, but then add an allowUnweighted flag in BlossomVOptions (default false) to allow the exception to be suppressed.  That way, an unaware developer will encounter the exception, and then decide whether they really want to go ahead with using Blossom V for this case; if so, they can pass a non-default option.